File: Sources/Services/PanelSnapper.swift
import CoreGraphics
import Foundation

/// PanelSnapper: Finds the exact polygon boundary of a comic panel by scanning
/// **inward from outside** the panel's approximate boundary for each row.
///
/// For each row y in the panel's vertical range:
///   1. Interpolate where the initial polygon's left and right edges are at row y.
///   2. Start slightly OUTSIDE those edges (in the white gutter).
///   3. Scan inward until the first non-gutter pixel — that's the panel border.
///
/// This approach is immune to white text, speech bubbles, and bright highlights
/// inside the panel, because we only ever scan through the gutter, never through
/// the panel artwork itself.
///
/// Slanted panels are handled naturally: the per-row right/left edges trace the
/// diagonal, and top-band vs bottom-band averages give the 4 polygon corners.
public class PanelSnapper {

    // MARK: - Public API

    public static func snapPoints(
        _ points: [CGPoint],
        in cgImage: CGImage
    ) -> [CGPoint] {
        guard points.count == 4 else { return points }

        let W = cgImage.width, H = cgImage.height
        guard W > 4, H > 4 else { return points }

        let bpp = 4, bpr = bpp * W
        var raw = [UInt8](repeating: 0, count: H * bpr)

        guard let ctx = CGContext(
            data: &raw, width: W, height: H,
            bitsPerComponent: 8, bytesPerRow: bpr,
            space: CGColorSpaceCreateDeviceRGB(),
            bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
        ) else { return points }

        // Core Graphics has Y=0 at the BOTTOM by default.
        // Apply a flip transform so the image is drawn top-down into the buffer:
        //   buffer row 0 = visual top of image (ny=0)
        //   buffer row H-1 = visual bottom of image (ny=1)
        // Without this, all y-coordinate lookups read the wrong rows (upside-down).
        ctx.draw(cgImage, in: CGRect(x: 0, y: 0, width: W, height: H))

        // ── 1. Determine if background is dark or light, and define gutter test ─
        let isDarkBg = isPageBackgroundDark(raw: raw, W: W, H: H, bpr: bpr, bpp: bpp)

        func isGutter(_ x: Int, _ y: Int) -> Bool {
            guard x >= 0, x < W, y >= 0, y < H else { return true }
            let o = y * bpr + x * bpp
            let r = Double(raw[o])
            let g = Double(raw[o+1])
            let b = Double(raw[o+2])
            
            if isDarkBg {
                let tol = 50.0
                return r < tol && g < tol && b < tol
            } else {
                let threshold = 205.0
                return r > threshold && g > threshold && b > threshold
            }
        }

        // ── 2. Polygon corners in pixel space ─────────────────────────────────
        // Input order: [TL, TR, BR, BL]
        let tl = CGPoint(x: points[0].x * Double(W), y: points[0].y * Double(H))
        let tr = CGPoint(x: points[1].x * Double(W), y: points[1].y * Double(H))
        let br = CGPoint(x: points[2].x * Double(W), y: points[2].y * Double(H))
        let bl = CGPoint(x: points[3].x * Double(W), y: points[3].y * Double(H))

        let topY = Int(min(tl.y, tr.y))
        let botY = Int(max(bl.y, br.y))
        guard topY < botY else { return points }

        // ── 3. Linear interpolation along polygon left/right edges ────────────
        // Left  edge: TL→BL   Right edge: TR→BR
        func lerpX(_ y: Int, from p0: CGPoint, to p1: CGPoint) -> Int {
            guard p1.y > p0.y else { return Int(p0.x) }
            let t = (Double(y) - p0.y) / (p1.y - p0.y)
            return Int(p0.x + t * (p1.x - p0.x))
        }

        // ── 4. Per-row scan: find left and right artwork edges using transition-based matching ─
        // We look for the exact transition between gutter and artwork in a narrow search window.
        // This is extremely robust and prevents drifting or merging into adjacent panels.
        let scanWidth = max(15, Int(Double(W) * 0.02)) // 2% of width as search window
        let scanHeight = max(15, Int(Double(H) * 0.02)) // 2% of height as search window

        var rowLeftEdge  = [Int](repeating: -1, count: H)
        var rowRightEdge = [Int](repeating: -1, count: H)

        for y in topY...botY {
            let polyLeft  = lerpX(y, from: tl, to: bl)
            let polyRight = lerpX(y, from: tr, to: br)

            // --- Left edge: find transition from gutter (true) to artwork (false) ---
            let scanLStart = max(0, polyLeft - scanWidth)
            let scanLEnd   = min(W - 1, polyLeft + scanWidth)
            var leftFound  = polyLeft

            var bestLDist = Double.infinity
            for x in scanLStart..<scanLEnd {
                if isGutter(x, y) && !isGutter(x + 1, y) {
                    let dist = abs(Double(x + 1) - Double(polyLeft))
                    if dist < bestLDist {
                        bestLDist = dist
                        leftFound = x + 1
                    }
                }
            }
            rowLeftEdge[y] = leftFound

            // --- Right edge: find transition from artwork (false) to gutter (true) ---
            let scanRStart = max(0, polyRight - scanWidth)
            let scanREnd   = min(W - 1, polyRight + scanWidth)
            var rightFound = polyRight

            var bestRDist = Double.infinity
            for x in scanRStart..<scanREnd {
                if !isGutter(x, y) && isGutter(x + 1, y) {
                    let dist = abs(Double(x) - Double(polyRight))
                    if dist < bestRDist {
                        bestRDist = dist
                        rightFound = x
                    }
                }
            }
            rowRightEdge[y] = rightFound
        }

        // ── 5. Per-column scan: find top and bottom artwork edges ──────────────

        func topBot(col x: Int) -> (top: Int, bot: Int) {
            // Top: find transition from gutter (true) to artwork (false)
            let scanTStart = max(0, topY - scanHeight)
            let scanTEnd   = min(H - 1, topY + scanHeight)
            var topFound   = topY

            var bestTDist = Double.infinity
            for y in scanTStart..<scanTEnd {
                if isGutter(x, y) && !isGutter(x, y + 1) {
                    let dist = abs(Double(y + 1) - Double(topY))
                    if dist < bestTDist {
                        bestTDist = dist
                        topFound = y + 1
                    }
                }
            }

            // Bottom: find transition from artwork (false) to gutter (true)
            let scanBStart = max(0, botY - scanHeight)
            let scanBEnd   = min(H - 1, botY + scanHeight)
            var botFound   = botY

            var bestBDist = Double.infinity
            for y in scanBStart..<scanBEnd {
                if !isGutter(x, y) && isGutter(x, y + 1) {
                    let dist = abs(Double(y) - Double(botY))
                    if dist < bestBDist {
                        bestBDist = dist
                        botFound = y
                    }
                }
            }

            return (topFound, botFound)
        }

        // Sample top/bottom from a band of columns in the center of the panel
        let midCol   = (Int(tl.x) + Int(tr.x)) / 2
        let colBand  = max(2, (Int(tr.x) - Int(tl.x)) / 6)
        let colRange = max(0, midCol - colBand)...min(W - 1, midCol + colBand)

        var topEdges: [Int] = [], botEdges: [Int] = []
        for x in colRange {
            let (t, b) = topBot(col: x)
            topEdges.append(t); botEdges.append(b)
        }

        let finalTop = topEdges.min() ?? topY
        let finalBot = botEdges.max() ?? botY

        // ── 6. Derive 4 corners from per-row edge arrays ──────────────────────
        // Average top 12.5% of rows for TL/TR, bottom 12.5% for BL/BR
        let rowCount = botY - topY + 1
        let band     = max(1, rowCount / 8)

        func avgLeft(_ rows: any Sequence<Int>) -> Int {
            let vals = rows.filter { $0 >= 0 && $0 < H && rowLeftEdge[$0] >= 0 }.map { rowLeftEdge[$0] }
            return vals.isEmpty ? Int(tl.x) : vals.reduce(0, +) / vals.count
        }
        func avgRight(_ rows: any Sequence<Int>) -> Int {
            let vals = rows.filter { $0 >= 0 && $0 < H && rowRightEdge[$0] >= 0 }.map { rowRightEdge[$0] }
            return vals.isEmpty ? Int(tr.x) : vals.reduce(0, +) / vals.count
        }

        let topRows = (topY...(topY + band))
        let botRows = ((botY - band)...botY)

        let tlPx = avgLeft(topRows);   let trPx = avgRight(topRows)
        let blPx = avgLeft(botRows);   let brPx = avgRight(botRows)

        // ── 7. Normalize to [0,1] ─────────────────────────────────────────────
        func nx(_ v: Int) -> Double { max(0, min(1, Double(v) / Double(W))) }
        func ny(_ v: Int) -> Double { max(0, min(1, Double(v) / Double(H))) }

        return [
            CGPoint(x: nx(tlPx), y: ny(finalTop)),  // TL
            CGPoint(x: nx(trPx), y: ny(finalTop)),  // TR
            CGPoint(x: nx(brPx), y: ny(finalBot)),  // BR
            CGPoint(x: nx(blPx), y: ny(finalBot))   // BL
        ]
    }

    private static func isPageBackgroundDark(raw: [UInt8], W: Int, H: Int, bpr: Int, bpp: Int) -> Bool {
        var brightnesses: [Double] = []
        
        func sample(_ x: Int, _ y: Int) {
            guard x >= 0, x < W, y >= 0, y < H else { return }
            let o = y * bpr + x * bpp
            guard o + 2 < raw.count else { return }
            let r = Double(raw[o])
            let g = Double(raw[o+1])
            let b = Double(raw[o+2])
            brightnesses.append((r + g + b) / 3.0)
        }
        
        let xStep = max(1, W / 60), yStep = max(1, H / 60)
        for x in stride(from: 0, to: W, by: xStep) {
            for d in 0...4 { sample(x, d); sample(x, H-1-d) }
        }
        for y in stride(from: 0, to: H, by: yStep) {
            for d in 0...4 { sample(d, y); sample(W-1-d, y) }
        }
        
        guard !brightnesses.isEmpty else { return false }
        brightnesses.sort()
        return brightnesses[brightnesses.count / 2] < 80.0
    }
}

File: shared/src/commonMain/kotlin/com/paneling/services/PanelSnapper.kt
package com.paneling.services

import com.paneling.models.PointF

object PanelSnapper {

    fun snapPoints(
        points: List<PointF>,
        raw: ByteArray,
        W: Int,
        H: Int
    ): List<PointF> {
        if (points.size != 4) return points
        if (W <= 4 || H <= 4) return points

        val bpp = 4
        val bpr = bpp * W

        // 1. Determine if background is dark or light, and define gutter test
        val isDarkBg = isPageBackgroundDark(raw, W, H, bpr, bpp)

        fun isGutter(x: Int, y: Int): Boolean {
            if (x < 0 || x >= W || y < 0 || y >= H) return true
            val o = y * bpr + x * bpp
            if (o + 2 >= raw.size) return true
            val r = (raw[o].toInt() and 0xFF).toDouble()
            val g = (raw[o + 1].toInt() and 0xFF).toDouble()
            val b = (raw[o + 2].toInt() and 0xFF).toDouble()

            return if (isDarkBg) {
                val tol = 50.0
                r < tol && g < tol && b < tol
            } else {
                val threshold = 205.0
                r > threshold && g > threshold && b > threshold
            }
        }

        // 2. Polygon corners in pixel space
        // Input order: [TL, TR, BR, BL]
        val tl = PointF(points[0].x * W, points[0].y * H)
        val tr = PointF(points[1].x * W, points[1].y * H)
        val br = PointF(points[2].x * W, points[2].y * H)
        val bl = PointF(points[3].x * W, points[3].y * H)

        val topY = minOf(tl.y, tr.y).toInt()
        val botY = maxOf(bl.y, br.y).toInt()
        if (topY >= botY) return points

        // 3. Linear interpolation along polygon left/right edges
        // Left edge: TL->BL   Right edge: TR->BR
        fun lerpX(y: Int, p0: PointF, p1: PointF): Int {
            if (p1.y <= p0.y) return p0.x.toInt()
            val t = (y.toDouble() - p0.y) / (p1.y - p0.y)
            return (p0.x + t * (p1.x - p0.x)).toInt()
        }

        // 4. Per-row scan: find left and right artwork edges using transition-based matching
        val scanWidth = maxOf(15, (W * 0.02).toInt())
        val scanHeight = maxOf(15, (H * 0.02).toInt())

        val rowLeftEdge = IntArray(H) { -1 }
        val rowRightEdge = IntArray(H) { -1 }

        for (y in topY..botY) {
            if (y < 0 || y >= H) continue
            val polyLeft = lerpX(y, tl, bl)
            val polyRight = lerpX(y, tr, br)

            // --- Left edge: find transition from gutter (true) to artwork (false) ---
            val scanLStart = maxOf(0, polyLeft - scanWidth)
            val scanLEnd = minOf(W - 1, polyLeft + scanWidth)
            var leftFound = polyLeft

            var bestLDist = Double.POSITIVE_INFINITY
            for (x in scanLStart until scanLEnd) {
                if (isGutter(x, y) && !isGutter(x + 1, y)) {
                    val dist = Math.abs((x + 1).toDouble() - polyLeft)
                    if (dist < bestLDist) {
                        bestLDist = dist
                        leftFound = x + 1
                    }
                }
            }
            rowLeftEdge[y] = leftFound

            // --- Right edge: find transition from artwork (false) to gutter (true) ---
            val scanRStart = maxOf(0, polyRight - scanWidth)
            val scanREnd = minOf(W - 1, polyRight + scanWidth)
            var rightFound = polyRight

            var bestRDist = Double.POSITIVE_INFINITY
            for (x in scanRStart until scanREnd) {
                if (!isGutter(x, y) && isGutter(x + 1, y)) {
                    val dist = Math.abs(x.toDouble() - polyRight)
                    if (dist < bestRDist) {
                        bestRDist = dist
                        rightFound = x
                    }
                }
            }
            rowRightEdge[y] = rightFound
        }

        // 5. Per-column scan: find top and bottom artwork edges
        fun topBot(x: Int): Pair<Int, Int> {
            // Top: find transition from gutter (true) to artwork (false)
            val scanTStart = maxOf(0, topY - scanHeight)
            val scanTEnd = minOf(H - 1, topY + scanHeight)
            var topFound = topY

            var bestTDist = Double.POSITIVE_INFINITY
            for (y in scanTStart until scanTEnd) {
                if (isGutter(x, y) && !isGutter(x, y + 1)) {
                    val dist = Math.abs((y + 1).toDouble() - topY)
                    if (dist < bestTDist) {
                        bestTDist = dist
                        topFound = y + 1
                    }
                }
            }

            // Bottom: find transition from artwork (false) to gutter (true)
            val scanBStart = maxOf(0, botY - scanHeight)
            val scanBEnd = minOf(H - 1, botY + scanHeight)
            var botFound = botY

            var bestBDist = Double.POSITIVE_INFINITY
            for (y in scanBStart until scanBEnd) {
                if (!isGutter(x, y) && isGutter(x, y + 1)) {
                    val dist = Math.abs(y.toDouble() - botY)
                    if (dist < bestBDist) {
                        bestBDist = dist
                        botFound = y
                    }
                }
            }

            return Pair(topFound, botFound)
        }

        // Sample top/bottom from a band of columns in the center of the panel
        val midCol = (tl.x.toInt() + tr.x.toInt()) / 2
        val colBand = maxOf(2, (tr.x.toInt() - tl.x.toInt()) / 6)
        val colRange = maxOf(0, midCol - colBand)..minOf(W - 1, midCol + colBand)

        val topEdges = mutableListOf<Int>()
        val botEdges = mutableListOf<Int>()
        for (x in colRange) {
            val (t, b) = topBot(x)
            topEdges.add(t)
            botEdges.add(b)
        }

        val finalTop = topEdges.minOrNull() ?: topY
        val finalBot = botEdges.maxOrNull() ?: botY

        // 6. Derive 4 corners from per-row edge arrays
        val rowCount = botY - topY + 1
        val band = maxOf(1, rowCount / 8)

        fun avgLeft(rows: Iterable<Int>): Int {
            val vals = rows.filter { it in 0 until H && rowLeftEdge[it] >= 0 }.map { rowLeftEdge[it] }
            return if (vals.isEmpty()) tl.x.toInt() else vals.sum() / vals.size
        }
        fun avgRight(rows: Iterable<Int>): Int {
            val vals = rows.filter { it in 0 until H && rowRightEdge[it] >= 0 }.map { rowRightEdge[it] }
            return if (vals.isEmpty()) tr.x.toInt() else vals.sum() / vals.size
        }

        val topRows = topY..(topY + band)
        val botRows = (botY - band)..botY

        val tlPx = avgLeft(topRows)
        val trPx = avgRight(topRows)
        val blPx = avgLeft(botRows)
        val brPx = avgRight(botRows)

        // 7. Normalize to [0,1]
        fun nx(v: Int): Double {
            return maxOf(0.0, minOf(1.0, v.toDouble() / W))
        }
        fun ny(v: Int): Double {
            return maxOf(0.0, minOf(1.0, v.toDouble() / H))
        }

        return listOf(
            PointF(nx(tlPx), ny(finalTop)), // TL
            PointF(nx(trPx), ny(finalTop)), // TR
            PointF(nx(brPx), ny(finalBot)), // BR
            PointF(nx(blPx), ny(finalBot))  // BL
        )
    }

    private fun isPageBackgroundDark(raw: ByteArray, W: Int, H: Int, bpr: Int, bpp: Int): Boolean {
        val brightnesses = mutableListOf<Double>()

        fun sample(x: Int, y: Int) {
            if (x in 0 until W && y in 0 until H) {
                val o = y * bpr + x * bpp
                if (o + 2 < raw.size) {
                    val r = (raw[o].toInt() and 0xFF).toDouble()
                    val g = (raw[o + 1].toInt() and 0xFF).toDouble()
                    val b = (raw[o + 2].toInt() and 0xFF).toDouble()
                    brightnesses.add((r + g + b) / 3.0)
                }
            }
        }

        val xStep = maxOf(1, W / 60)
        val yStep = maxOf(1, H / 60)
        for (x in 0 until W step xStep) {
            for (d in 0..4) {
                sample(x, d)
                sample(x, H - 1 - d)
            }
        }
        for (y in 0 until H step yStep) {
            for (d in 0..4) {
                sample(d, y)
                sample(W - 1 - d, y)
            }
        }

        if (brightnesses.isEmpty()) return false
        brightnesses.sort()
        return brightnesses[brightnesses.size / 2] < 80.0
    }
}
